src/xdsl/pattern_rewriter.py

```diff
@@ -0,0 +1,197 @@
+            "have two non-self arguments.")
+    is_method = params[0].name == "self"
+    if is_method:
+        if len(params) != 3:
+            raise Exception(
+                "op_type_rewrite_pattern expects the decorated method to "
+                "have two non-self arguments.")
+    else:
+        if len(params) != 2:
+            raise Exception(
+                "op_type_rewrite_pattern expects the decorated function to "
+                "have two arguments.")
+    expected_type = params[-2].annotation
+    if not issubclass(expected_type, Operation):
+        raise Exception(
+            "op_type_rewrite_pattern expects the first non-self argument"
+            "type hint to be an Operation subclass")
+
+    if not is_method:
+
+        def op_type_rewrite_pattern_static_wrapper(
+                op: Operation, rewriter: PatternRewriter) -> None:
+            if not isinstance(op, expected_type):
+                return None
+            func(op, rewriter)
+
+        return op_type_rewrite_pattern_static_wrapper
+
+    def op_type_rewrite_pattern_method_wrapper(
+            self, op: Operation, rewriter: PatternRewriter) -> None:
+        if not isinstance(op, expected_type):
+            return None
+        func(self, op, rewriter)
+
+    return op_type_rewrite_pattern_method_wrapper
+
+
+@dataclass(eq=False, repr=False)
+class GreedyRewritePatternApplier(RewritePattern):
+    """Apply a list of patterns in order until one pattern matches, and then use this rewrite."""
+
+    rewrite_patterns: List[RewritePattern]
+    """The list of rewrites to apply in order."""
+
+    def match_and_rewrite(self, op: Operation,
+                          rewriter: PatternRewriter) -> None:
+        for pattern in self.rewrite_patterns:
+            pattern.match_and_rewrite(op, rewriter)
+            if rewriter.has_done_action:
+                return
+        return
+
+
+@dataclass(eq=False, repr=False)
+class PatternRewriteWalker:
+    """
+    Walks the IR in the block and instruction order, and rewrite it in place.
+    Previous references to the walked operations are invalid after the walk.
+    Can walk either first the regions, or first the owner operation.
+    The walker will also walk recursively on the created operations.
+    """
+
+    pattern: RewritePattern
+    """Pattern to apply during the walk."""
+
+    walk_regions_first: bool = field(default=False)
+    """Choose if the walker should first walk the operation regions first, or the operation itself."""
+
+    apply_recursively: bool = field(default=True)
+    """Apply recursively rewrites on new operations."""
+
+    walk_reverse: bool = field(default=False)
+    """Walk the regions and blocks in reverse order. That way, all uses are replaced before the definitions."""
+
+    def rewrite_module(self, op: ModuleOp):
+        """Rewrite an entire module operation."""
+        self._rewrite_op(op)
+
+    def _rewrite_op(self, op: Operation) -> int:
```

Comment:
I think the return value here should have a comment what it means. AFAIU, it is the change in the index to the next operation, so if you add multiple you have to "jump" further, which feels like an unnecessary inter-dependency between `_rewrite_op` and `_rewrite_op_regions`. Could this  just return the number of added operations in any case, or am I missing something?

Reply:
This is a bit more complex than just adding the values, since you have a difference when you are walking forward and backwards, and if you are rewriting recursively or not.

Reply:
Anyway I plan to remove this function, or change it a lot because this will get more complex once you have rewritings of blocks and regions.